File: sem2/po/po_proj2/src/studia/prog_ob/projekt2/world_simulator/organisms/plants/Plant.java
package studia.prog_ob.projekt2.world_simulator.organisms.plants;

import studia.prog_ob.projekt2.world_simulator.RandomGenerator;
import studia.prog_ob.projekt2.world_simulator.SimLogger;
import studia.prog_ob.projekt2.world_simulator.World;
import studia.prog_ob.projekt2.world_simulator.maps.MapLocation;
import studia.prog_ob.projekt2.world_simulator.organisms.Organism;

public abstract class Plant extends Organism {
    public static final int SEEDING_CHANCE_BASE = 100;
    public static final int BASE_PLANT_SEEDING_THRESHOLD = 97;
    public static final int BASE_PLANT_STRENGTH = 0;
    public static final int BASE_PLANT_INITIATIVE = 0;


    private final int seeding_threshold;
    public Plant(int new_strength, int new_initiative, MapLocation new_map_location, char new_symbol, World new_world, int new_seeding_threshold) {
        super(new_strength, new_initiative, new_map_location, new_symbol, new_world);
        this.seeding_threshold = new_seeding_threshold;
    }


    @Override
    public void action() {
        seed();
    }

    @Override
    public void collision(Organism invader) {
        invader.update_location(getMap_location());
        kill_self();
    }

    protected Plant seed() {
        int r = RandomGenerator.rand_int(SEEDING_CHANCE_BASE);

        if (r >= seeding_threshold) {
            for (int i = -1; i < 2; i++) {
                for (int j = -1; j < 2; j++) {
                    if (i == 0 && j == 0)
                        continue;

                    MapLocation l = new MapLocation(getMap_location());
                    l.x += j;
                    l.y += i;
                    if(!getWorld().getMap().out_of_map(l) && getWorld().getMap().get_object(l) == null) {
                        SimLogger.log(this + SimLogger.SEEDING_MESSAGE);
                        return (Plant) getWorld().add_organism(spawn_offspring(l));
                    }
                }
            }
        } else {
            //SimLogger.log(this + SimLogger.PLANT_IDLE_MESSAGE);
        }

        return null;
    }

    @Override
    public String toString() {
        return "Plant " + getMap_location() + " ";
    }
}

File: sem2/po/po_proj3/Plant.py
import random

from Organism import Organism

SEEDING_CHANCE_BASE = 100
BASE_PLANT_SEEDING_THRESHOLD = 97
BASE_PLANT_STRENGTH = 0
BASE_PLANT_INITIATIVE = 0


class Plant(Organism):
    def __init__(self, new_world, new_location, new_strength, new_initiative, new_type, new_seeding_threshold):
        super().__init__(new_world, new_location, new_strength, new_initiative, new_type)
        self._seeding_threshold = new_seeding_threshold

    def action(self):
        self.seed()

    def collision(self, invader):
        invader.update_location(self._location)
        self.kill_self()

    def seed(self):
        r = random.randint(0, SEEDING_CHANCE_BASE)

        if r > self._seeding_threshold:
            free_loc = self._world.map.get_first_free_neighbour(self._location)
            if free_loc is not None:
                self.tlog(" seeded at " + str(free_loc))
                self._world.add_organism(self.spawn_offspring(free_loc))

    def __str__(self):
        return "Plant " + str(self._location)